drivers: retained_mem: add BBRAM driver

rbrscoleman opened the pull request:
Adds a non-initialised BBRAM-based retained memory driver.

Based on the RAM-based retained memory driver (`zephyr,retained-ram`).

This is convenient for storing the boot mode in a BBRAM device, such as the `st,stm32-bbram` in many STM32 RTCs:

```dts
/ {
	chosen {
		zephyr,boot-mode = &boot_mode0;
	};
};

&bbram {
	status = "okay";

	retainedmem0: retainedmem {
		compatible = "zephyr,retained-bbram";
		status = "okay";

		boot_mode0: boot_mode@0 {
			compatible = "zephyr,retention";
			status = "okay";
			reg = <0x0 0x1>;
		};
	};
};
```

In addition to native sim, I tested on two NUCLEOs I had close at hand (plus, we've been using this internally on our own hardware for a while). I don't have access to any non-STM32 BBRAM hardware to test this on.

tpambor commented:
Currently, the documentation for the retention system states that the data should not be stored to a battery-backed RAM.
https://github.com/zephyrproject-rtos/zephyr/blob/84de5b841713a0da97986e1edb3319edb354f904/doc/services/retention/index.rst?plain=1#L10-L13

If it is ok to do so, the documentation should be updated.

yperess commented:
Sorry for the delay, went through a career change and doing Zephyr on my own time now. I'm happy for BBRAM to merge other memory APIs if someone wants to make the migration.

yperess commented on drivers/retained_mem/retained_mem_zephyr_bbram.c:
```diff
@@ -0,0 +1,164 @@
+
+	err = bbram_get_size(config->bbram, &total_size);
+	if (err < 0) {
+		return err;
```
You should either unlock here or use a goto and have a single exit point for the function.

nordicjm commented on tests/drivers/retained_mem/api/boards/native_sim_bbram.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_BBRAM=y
```
put this into `api/prj_bbram.conf` instead

nordicjm reviewed (commented):
Like the pre-kernel check part, might add that to the other drivers in a future PR

rbrscoleman committed:
drivers: retained_mem: forward mutex init error

Previously, mutex initialization failed quietly across all retained memory
driver implementations.

Signed-off-by: Samuel Coleman <samuel.coleman@rbr-global.com>

rbrscoleman committed:
tests: drivers: retained_mem: fix compilation

Fixes the compilation error:

    error: 'static' is not at beginning of declaration
    [-Werror=old-style-declaration]

Signed-off-by: Samuel Coleman <samuel.coleman@rbr-global.com>

rbrscoleman committed:
drivers: retained_mem: add BBRAM driver

Adds a non-initialised BBRAM-based retained memory driver.

Based on the RAM-based retained memory driver (`zephyr,retained-ram`).

Signed-off-by: Samuel Coleman <samuel.coleman@rbr-global.com>

rbrscoleman committed:
tests: drivers: retained_mem: add BBRAM test

Tests the BBRAM-backed retained memory driver on the native simulator and
on ST NUCLEO-L476RE and NUCLEO-L452RE.

Signed-off-by: Samuel Coleman <samuel.coleman@rbr-global.com>
